listen on all interfaces by default, sane -gossip behaviour

tbg opened the pull request:
- passing `-gossip` is now required, but you can pass `self://`.
  this does not happen in the resolver logic itself yet, but I suppose
  it will be something that practically refactors itself as the resolver
  logic gets more sophisticated.
- -addr :port listens on all interfaces.
- updated the README and flags (and made it more copy-paste friendly)

vivekmenezes commented on server/server_test.go:
```diff
@@ -112,7 +112,8 @@ func TestInitEngine(t *testing.T) {
 	for _, spec := range testCases {
-		ctx := &Context{Stores: spec.key}
+		ctx := NewContext()
+		ctx.Stores, ctx.GossipBootstrap = spec.key, "self://"
```
ctx := &Context{Stores: spec.key, GossipBootstrap: "self://"}
would work right?

tschottdorf replied:
yes, that currently does pretty much the same.

bdarnell replied:
`NewContext()` sets some default values that wouldn't be present with `&Context{}`.

bdarnell commented on server/context.go:
```diff
@@ -165,28 +170,23 @@ func (ctx *Context) initEngine(attrsStr, path string) (engine.Engine, error) {
+		// uniquely for use in Gossip. This avoids having to specify
+		// the port for single-node clusters twice (once in -addr,
+		// once in -gossip).
+		if strings.Contains(address, "self://") {
```
This looks like it should be an exact match instead of Contains().

bdarnell commented:
LGTM

tbg committed:
listen on all interfaces by default, sane -gossip behaviour

* passing `-gossip` is now required, but you can pass `self://`.
  this does not happen in the resolver logic itself yet, but I suppose
  it will be something that practically refactors itself as the resolver
  logic gets more sophisticated.
* -addr :port listens on all interfaces.
* updated the README and flags.

spencerkimball commented on server/context.go:
```diff
@@ -139,10 +141,13 @@ func (ctx *Context) Init() error {
 	ctx.NodeAttributes = parseAttributes(ctx.Attrs)
 
-	resolvers, err := parseGossipBootstrapAddrs(ctx.GossipBootstrap)
+	resolvers, err := ctx.parseGossipBootstrapAddrs()
```
Let's rename this method to `parseGossipBootstrapResolvers()`

tschottdorf replied:
done.

spencerkimball commented on server/context.go:
```diff
@@ -139,10 +141,13 @@ func (ctx *Context) Init() error {
 		return err
 	}
+	if len(resolvers) == 0 {
+		return errors.New("no resolvers specified for gossip network, did you pass -gossip?")
```
s/pass/specify/

tschottdorf replied:
done.

spencerkimball commented on server/context.go:
```diff
@@ -165,28 +170,23 @@ func (ctx *Context) initEngine(attrsStr, path string) (engine.Engine, error) {
-func parseGossipBootstrapAddrs(gossipBootstrap string) ([]*gossip.Resolver, error) {
+func (ctx *Context) parseGossipBootstrapAddrs() ([]*gossip.Resolver, error) {
 	var bootstrapResolvers []*gossip.Resolver
+	gossipBootstrap := ctx.GossipBootstrap
```
Why create a stack variable for gossipBootstrap?

tschottdorf replied:
done.

spencerkimball commented on server/server.go:
```diff
@@ -75,8 +75,7 @@ func NewServer(ctx *Context, stopper *util.Stopper) (*Server, error) {
 
-	// If the specified address includes no host component, use the hostname.
-	addr := util.EnsureHost(ctx.Addr)
+	addr := ctx.Addr
```
This is all well and good but if you pass in :8080 to the rpc server, what does it report from `Addr()` after binding to all network interfaces? That address is what we gossip with the node descriptor so that other nodes can connect.

tschottdorf replied:
see the next PR.

tschottdorf commented:
@spencerkimball PR coming up soon for your comments.
